app/models/settings/base.rb

```diff
@@ -0,0 +1,155 @@
+
+    class << self
+      def clear_cache
+        RequestStore.delete(cache_key)
+        Rails.cache.delete(cache_key)
+      end
+
+      def field(key, default: nil, type: :string, separator: nil, validates: nil)
+        define_field(
+          key,
+          default: default,
+          type: type,
+          separator: separator,
+          validates: validates,
+        )
+      end
+
+      def get_field(key)
+        @defined_fields.detect { |field| field[:key] == key.to_s } || {}
+      end
+
+      def cache_key
+        @cache_key ||= name.underscore
+      end
+
+      def keys
+        @defined_fields.pluck(:key)
+      end
+
+      private
+
+      def define_field(key, default: nil, type: :string, separator: nil, validates: nil)
+        key = key.to_s
+
+        raise(ProcetedKeyError, "Can't use #{key} as setting key.") if key.in?(PROTECTED_KEYS)
+
+        @defined_fields ||= []
+        @defined_fields << {
+          key: key,
+          default: default,
+          type: type || :string
+        }
+
+        # Getter
+        define_singleton_method(key) do
+          result = __send__(:value_of, key)
+          result ||= default.is_a?(Proc) ? default.call : default
+          result = __send__(:convert_string_to_value_type, type, result, separator: separator)
+
+          result
+        end
+
+        # Setter
+        define_singleton_method("#{key}=") do |value|
+          var_name = key
+
+          record = find_by(var: var_name) || new(var: var_name)
+          value = __send__(:convert_string_to_value_type, type, value, separator: separator)
+
+          record.value = value
+          record.save!
+
+          value
+        end
+
+        # Validation
+        if validates
+          validates[:if] = proc { |item| item.var.to_s == key }
+          __send__(:validates, key, **validates)
+
+          define_method(:read_attribute_for_validation) { |_key| self.value }
+        end
+
+        return unless type == :boolean
+
+        # Predicate method for booleans
+        define_singleton_method("#{key}?") { __send__(key) }
+      end
+
+      def convert_string_to_value_type(type, value, separator: nil)
+        return value unless value.class.in?([String, Integer, Float, BigDecimal])
+
+        case type
+        when :boolean
+          value.in?(["true", "1", 1, true])
+        when :array
+          value.split(separator || SEPARATOR_REGEXP).reject(&:empty?).map(&:strip)
+        when :hash
+          value = begin
+            YAML.safe_load(value).to_h
+          rescue StandardError
+            {}
+          end
+          value.deep_stringify_keys!
+          ActiveSupport::HashWithIndifferentAccess.new(value)
+        when :integer
+          value.to_i
+        when :float
+          value.to_f
+        when :big_decimal
+          value.to_d
+        else
+          value
+        end
+      end
+
+      def value_of(var_name)
+        unless Database.table_available?(table_name)
```

Comment:
We already had this code in place so I removed the custom version the gem was using.